File: new_ui/lib/pricing.ts
export interface PricingConfig {
  base_rate_per_min_cents: number
  fix_timing_addon_cents: number
  segment_regen_cost_cents: number
  volume_discounts: { min_minutes: number; discount_pct: number }[]
}

export function calculateCost(
  durationMinutes: number,
  options: { fixTiming: boolean },
  pricing: PricingConfig
): number {
  const rate = pricing.base_rate_per_min_cents + (options.fixTiming ? pricing.fix_timing_addon_cents : 0)
  const subtotal = durationMinutes * rate

  const discount_pct =
    pricing.volume_discounts
      .filter((t) => durationMinutes >= t.min_minutes)
      .sort((a, b) => b.min_minutes - a.min_minutes)[0]?.discount_pct ?? 0

  return Math.round(subtotal * (1 - discount_pct / 100))
}

export function calculateRegenCost(
  changedSegments: { startTime: number; endTime: number }[],
  pricing: PricingConfig,
  ttsMode?: "natural" | "original_voice",
  videoDurationMinutes?: number | null
): number {
  if (ttsMode === "original_voice") {
    // Flat cost = full video duration × rate/4, regardless of how many segments changed
    const mins = videoDurationMinutes ?? 0
    return Math.round(mins * pricing.base_rate_per_min_cents / 4)
  }
  const totalMinutes = changedSegments.reduce((sum, s) => sum + (s.endTime - s.startTime), 0) / 60
  return Math.round(totalMinutes * pricing.base_rate_per_min_cents)
}

export function formatCents(cents: number): string {
  return `$${(cents / 100).toFixed(2)}`
}

File: api/pricing.py
import math

PRICING = {
    "base_rate_per_min_cents": 100,   # $1.00 / min
    "fix_timing_addon_cents":   20,   # $0.20 / min extra
    "segment_regen_cost_cents":  3,   # $0.03 / segment
    "volume_discounts": [
        {"min_minutes":  5, "discount_pct": 10},
        {"min_minutes": 10, "discount_pct": 15},
    ],
}


def calculate_run_cost_cents(duration_minutes: float, fix_timing: bool) -> int:
    if duration_minutes <= 0:
        return 0
    rate = PRICING["base_rate_per_min_cents"]
    if fix_timing:
        rate += PRICING["fix_timing_addon_cents"]
    subtotal = duration_minutes * rate
    best_discount = 0
    for tier in sorted(PRICING["volume_discounts"], key=lambda t: t["min_minutes"], reverse=True):
        if duration_minutes >= tier["min_minutes"]:
            best_discount = tier["discount_pct"]
            break
    return round(subtotal * (1 - best_discount / 100))


def calculate_regen_cost_cents(duration_minutes: float, ttsmodel: int = 1) -> int:
    if duration_minutes <= 0:
        return 0
    # ttsmodel 3 = Original Voice — cheaper (1/4 rate); others = Natural Voice — full rate
    if ttsmodel == 3:
        return round(duration_minutes * PRICING["base_rate_per_min_cents"] / 4)
    return round(duration_minutes * PRICING["base_rate_per_min_cents"])
